### DISCUSSION TITLE
Integration into ESP-IDF

### DISCUSSION THREAD

**Discussion by KiShiVi** [pschatzmann/arduino-audio-tools#1600] [Q&A] [answered]
Good afternoon, could you please illustrate a clear way to integrate your library into an ESP-IDF project. I have seen your instructions for connecting it, but I get a number of problems following your steps. For example, when building a project, CMake fails on the "project(arduino-audio-tools)" command. I have tried to troubleshoot the problems, but the further I go, the more problems I have when building. I am very much looking forward to your help! Thanks!

**Comment by pschatzmann** [comment, marked as answer]
What are the exact error messages ?
I committed some corrections, but I noticed that the -DESP32_CMAKE is only passed properly if it is part of the platformio.ino build_flags =  -DESP32_CMAKE

**Reply by KiShiVi** [reply]
Most likely it's because I'm not setting up the CMake files correctly. I still don't understand how it is possible to add a component to a project, which is a project in itself (it has project() in Cmake on the 4th line).

Could you please describe in more detail how to configure this project as a component in your project (which CMake and how to fix it) for correct assembly.

the current structure of the project is as follows:
-components
---arduino-audio-tools
-----src
-----...
-----CMakeLists.txt
---ESP32-A2DP
-----src
-----...
-----CMakeLists.txt
-main
---CMakeLists.txt
---main.cpp
CMakeLists.txt

**Comment by pschatzmann** [comment]
Did you read [this already](https://github.com/pschatzmann/ESP32-A2DP/wiki/Espressif-IDF-as-a-Component) ? 
The AudioTools library is not a IDF component but a header only library. 
So all is needed to to is to inform your project about the location of the header files

**Reply by pschatzmann** [reply]
I committed a change which should support the AudioTools to be used as IDF component.
The CMakeList.txt in the root should contains the following line
idf_build_set_property(COMPILE_OPTIONS "-DESP32_CMAKE" APPEND)
or alternatively you can use the build_flags mentioned above...

**Comment by crosswick** [comment]
@pschatzmann first off thank you very much these audio tools, they look very useful. I am also looking into using them with ESP-IDF, preferably the most recent stable version which at the moment is v5.3 - I recall reading somewhere that for now, v5.1 worked better with AAT.

I’ve tried following your comments from June 20 as follows:

- look at https://github.com/pschatzmann/ESP32-A2DP/wiki/Espressif-IDF-as-a-Component for a general recipe for adding a component to IDF. 
- I ran into: the `idf.py create_project` command failed because of a missing CMakeLists.txt
- so I decided to try first installing the ESP-IDF extension in VS Code (cursor.com) and working from a vanilla/blink project created from there
- so from a basic project:
  - mkdir components
  - cd components
  - git clone https://github.com/pschatzmann/arduino-audio-tools

without any CMakeLists.txt edits this did not build:

```
CMake Warning at /Users/steven/esp/v5.1.4/esp-idf/tools/cmake/component.cmake:293 (message):
  No source files found for SRC_DIRS entry 'src'.
Call Stack (most recent call first):
  /Users/steven/esp/v5.1.4/esp-idf/tools/cmake/component.cmake:443 (__component_add_sources)
  components/arduino-audio-tools/CMakeLists.txt:6 (idf_component_register)

CMake Error at components/arduino-audio-tools/CMakeLists.txt:12 (target_compile_options):
  target_compile_options may only set INTERFACE properties on INTERFACE
  targets
```

I think this may be because there is still a /src directory in the git-cloned AAT which should be ignored but isn’t?

anyway, I proceded to do:

- rename /main/blink_example_main.c to /main/blink_example_main.cpp
- copy over (replace) the code inside of it from this Use Without Arduino example: https://github.com/pschatzmann/arduino-audio-tools/wiki/Using-the-Framework-w-o-Arduino#idf

which unfortunately builds even less:

```
 *  Executing task: ninja  

ninja: error: loading 'build.ninja': No such file or directory

 *  The terminal process "ninja" failed to launch (exit code: 1). 
```

I’ve also done some attempts yesterday with PlatformIO; I couldn’t get it to build there either, possibly having to do with -DESP32_CMAKE. I could detail my process with that too if you’d like - I don’t have a preference yet for ‘pure’ ESP-IDF in VS Code or via PlatformIO, and am open to advice.

Could you evaluate my approach please? Thank you in advance.

**Reply by pschatzmann** [reply]
It seems that we are getting some errors now because this is a header only library and therefore does not have any (c or ccp) source files.

I committed [some corrections](https://github.com/pschatzmann/arduino-audio-tools/commit/e5c04b5c447b9a7822e5382f42d3eb6d45d996ca)

**Reply by crosswick** [reply]
thank you - unfortunately I'm not getting it to work yet though. my steps just now:

- create new ESP-IDF v5.3 project in VS Code using the 'blink' template, which builds and flashes correctly
- mkdir components, cd components, git clone https://github.com/pschatzmann/arduino-audio-tools.git, still builds
- rename blink_example_main.c to blink_main_example.cpp
- replace its contents with the "w/o Arduino" IDF example code

the existing CMakeLists.txt in /main reads:
```
idf_component_register(SRCS "blink_example_main.c"
                       INCLUDE_DIRS ".")
```
I replaced this with:
```
FILE(GLOB_RECURSE app_sources ${CMAKE_SOURCE_DIR}/src/*.*)

idf_component_register(SRCS ${app_sources})
```
- ESP-IDF: Full Clean
- ESP-IDF: Build Project

which yields this error:

```
/Users/steven/esp/v5.3/esp-idf/components/freertos/app_startup.c:199:(.text.main_task+0x99): undefined reference to `app_main'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

 *  The terminal process "ninja" terminated with exit code: 1. 
```

**Comment by pschatzmann** [comment]
You are getting a linker error because app_main needs to be a c function but you defined it as using c++ calling convention!
Have another look at the documented example!

**Reply by crosswick** [reply]
...doesn't this IDF example code I copied already define app_main to be a c function? or do I misunderstand still
https://github.com/pschatzmann/arduino-audio-tools/wiki/Using-the-Framework-w-o-Arduino#idf

with

```
extern "C" void app_main() {
    setup();
    while(true) loop();
}
```

**Reply by pschatzmann** [reply]
yes - this should work...

Did you clean and rebuild the probject ?

**Reply by crosswick** [reply]
Yes, like I mentioned above. Just tried again, same result... anything I can check/verify/try?

**Reply by pschatzmann** [reply]
Not sure: I  started with an empty example and was getting:

```
Processing esp32dev (platform: espressif32; board: esp32dev; framework: espidf)
-----------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.6.0+sha.c9c9bbe) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-espidf @ 3.50201.0 (5.2.1) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.7.1 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0) 
 - toolchain-xtensa-esp32 @ 12.2.0+20230208
Reading CMake configuration...
Warning! Flash memory size mismatch detected. Expected 4MB, found 2MB!
Please select a proper value in your `sdkconfig.defaults` or via the `menuconfig` target!
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 0 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Retrieving maximum program size .pio/build/esp32dev/firmware.elf
Checking size .pio/build/esp32dev/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=         ]   5.6% (used 18352 bytes from 327680 bytes)
Flash: [=====     ]  48.3% (used 506385 bytes from 1048576 bytes)
======================= [SUCCESS] Took 30.11 seconds =======================
 *  Terminal will be reused by tasks, press any key to close it. 
 
```

**Comment by pschatzmann** [comment]
platformio.ini is

[env:esp32dev]
platform = espressif32
board = esp32dev
framework = espidf

**Reply by crosswick** [reply]
Ah, you are using PlatformIO - I am using the ESP-IDF extension in VS Code

**Reply by pschatzmann** [reply]
So, did you try w/o extern "C" ? 
I think this might have worked at some point of time...

**Reply by crosswick** [reply]
If you mean this, then yes, and it yields the same error:

```
//extern "C" void app_main() {
void app_main() {
    setup();
    while(true) loop();
}
```

same thing if I also rename blink_example_main.cpp to blink_example_main.c

perhaps there is something relevant to configure in ESP-IDF menuconfig?

meanwhile I'm also looking into the PlatformIO route

**Comment by crosswick** [comment]
Meanwhile using PlatformIO I'm getting these errors:

```
In file included from components/arduino-audio-tools/src/AudioAnalog/AnalogDriverBase.h:6,
                 from components/arduino-audio-tools/src/AudioAnalog/AnalogAudioStream.h:6,
                 from components/arduino-audio-tools/src/AudioTools.h:80,
                 from src/main.cpp:1:
components/arduino-audio-tools/src/AudioAnalog/AnalogConfigESP32V1.h:106:31: error: 'ADC_ATTEN_DB_12' was not declared in this scope; did you mean 'ADC_ATTEN_DB_11'?
```

```
In file included from components/arduino-audio-tools/src/AudioAnalog/AnalogAudioStream.h:7:
components/arduino-audio-tools/src/AudioAnalog/AnalogDriverESP32V1.h: In member function 'bool audio_tools::AnalogDriverESP32V1::setup_rx()':
components/arduino-audio-tools/src/AudioAnalog/AnalogDriverESP32V1.h:525:20: error: 'struct adc_continuous_handle_cfg_t' has no member named 'flags'
  525 |         adc_config.flags.flush_pool  = true;
```

by the way, did you not find that `build_flags = -DESP32_CMAKE` should go into platformio.ini, instead of this below in src/CMakeLists.txt?

```
# add arduino-audio-tools
add_compile_definitions(-DESP32_CMAKE )
```

with either or both in place I get the errors.

**Comment by pschatzmann** [comment]
It seems that you are using some old versions: Compare with the information I gave you above!

**Reply by crosswick** [reply]
I'm sorry, I don't see any version-related information in your remarks above. I am working with freshly downloaded/installed versions of the ESP-IDF and PlatformIO extensions and am doing simply `git clone https://github.com/pschatzmann/arduino-audio-tools.git`, so not using a previous version of that either... but I will re-check

**Reply by pschatzmann** [reply]
Just look at the log output from 1h ago

**Reply by crosswick** [reply]
ah, I see - sofar the only thing that I notice is
` - framework-espidf @ 3.50201.0 (5.2.1) `

is this not the version that PlatformIO automatically uses? I have not seen any instructions for specifying it if not

**Reply by pschatzmann** [reply]
I just kept Platfomio up to date...

**Reply by crosswick** [reply]
alright well I'm reinstalling PlatformIO on a different machine to see again if I can get a basic example working, let's see if anything changes

if I understand correctly everything should work with the most current versions; there's no need to specify older ones

**Reply by pschatzmann** [reply]
There is an Upgrade PlatfromIO Core function in the menu
And an Update function in the Platforms

**Reply by crosswick** [reply]
to compare, this is the start of the PlatformIO build output that failed for me earlier:

```
Processing esp32dev (platform: espressif32; board: esp32dev; framework: espidf)
-------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.4.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-espidf @ 3.50101.230828 (5.1.1) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.9.0 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0) 
 - toolchain-xtensa-esp32 @ 12.2.0+20230208
Reading CMake configuration...
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 0 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
```

**Reply by crosswick** [reply]
ah which uses IDF v5.1, so that was I guess found on that machine by PlatformIO instead of downloaded. let me update that

**Reply by crosswick** [reply]
ah after running the Update function in the Platforms I get an even newer version of IDF:

```
- framework-espidf @ 3.50300.0 (5.3.0) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.9.0 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 1.23800.240113 (2.38.0) 
 - toolchain-xtensa-esp-elf @ 13.2.0+20240530
```

so maybe it only works the the version you are running, v5.2.1? I'll try this v5.3 in PlatformIO first

**Reply by crosswick** [reply]
not working, no. 

with `add_compile_definitions(-DESP32_CMAKE )` in /src/CMakelists.txt, or `build_flags = -DESP32_CMAKE` in platformio.ini:

```
In file included from components/arduino-audio-tools/src/AudioConfig.h:31,
                 from components/arduino-audio-tools/src/AudioTools.h:66,
                 from src/main.cpp:1:
components/arduino-audio-tools/src/AudioLibs/Desktop/NoArduino.h:132:22: error: 'virtual void audio_tools::Stream::setTimeout(size_t)' was hidden [-Werror=overloaded-virtual=]
  132 |         virtual void setTimeout(size_t t){}
      |                      ^~~~~~~~~~
In file included from components/arduino-audio-tools/src/AudioTools/AudioStreams.h:9,
                 from components/arduino-audio-tools/src/AudioFilter/Equilizer.h:6,
                 from components/arduino-audio-tools/src/AudioTools.h:75:
components/arduino-audio-tools/src/AudioTools/BaseStream.h:221:8: note:   by 'void audio_tools::CatStream::setTimeout(uint32_t)'
  221 |   void setTimeout(uint32_t t) { _timeout = t; }
      |        ^~~~~~~~~~
Compiling .pio/build/esp32dev/esp_adc/adc_continuous.c.o
Compiling .pio/build/esp32dev/esp_adc/esp32/adc_dma.c.o
Compiling .pio/build/esp32dev/esp_adc/esp32/adc_cali_line_fitting.c.o
Compiling .pio/build/esp32dev/esp_adc/deprecated/esp32/esp_adc_cal_legacy.c.o
In file included from components/arduino-audio-tools/src/Concurrency/SynchronizedBuffers.h:4,
                 from components/arduino-audio-tools/src/AudioLibs/Concurrency.h:9,
                 from components/arduino-audio-tools/src/AudioTools.h:68:
components/arduino-audio-tools/src/AudioTools/AudioTypes.h:127:18: error: 'virtual void audio_tools::AudioInfo::logInfo(const char*)' was hidden [-Werror=overloaded-virtual=]
  127 |     virtual void logInfo(const char* source="") {
      |                  ^~~~~~~
In file included from components/arduino-audio-tools/src/AudioAnalog/AnalogDriverBase.h:6,
                 from components/arduino-audio-tools/src/AudioAnalog/AnalogAudioStream.h:6,
                 from components/arduino-audio-tools/src/AudioTools.h:80:
components/arduino-audio-tools/src/AudioAnalog/AnalogConfigESP32V1.h:131:10: note:   by 'void audio_tools::AnalogConfigESP32V1::logInfo()'
  131 |     void logInfo() {
```

I guess I could try to install ESP-IDF v5.2.1 and see if I get your results... thank you for your continued attention by the way, I appreciate it

**Reply by crosswick** [reply]
my first machine already has v5.2.1 installed, so that's not it. 

I do see a more recent version of ninja though! than you are running. perhaps that's the bottleneck?

```
Processing esp32dev (platform: espressif32; board: esp32dev; framework: espidf)
--------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.7.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-espidf @ 3.50201.240515 (5.2.1) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.9.0 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0) 
 - toolchain-xtensa-esp-elf @ 13.2.0+20230928
Reading CMake configuration...
Warning! Flash memory size mismatch detected. Expected 4MB, found 2MB!
Please select a proper value in your `sdkconfig.defaults` or via the `menuconfig` target!
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
```

same for my other machine, see above. GPT says ninja should get updated with PlatformIO Core

**Reply by pschatzmann** [reply]
These errors seem to come with the latest version. I am not sure yet how to deal with this best.
A current work around is to add the following line to platformio.ini

build_flags = -Wno-overloaded-virtual

**Reply by crosswick** [reply]
Alright, well I am not in a big hurry - I look forward to exploring the audio tools via PlatformIO-Arduino for now. Good to know at least that the issue seems to have been identified.

Unfortunately your workaround doesn't work for me, I get a long list of related errors, see the first three below:

```
Processing esp32dev (platform: espressif32; board: esp32dev; framework: espidf)
------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.7.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-espidf @ 3.50201.240515 (5.2.1) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.9.0 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0) 
 - toolchain-xtensa-esp-elf @ 13.2.0+20230928
Reading CMake configuration...
Generating assembly for certificate bundle...
Warning! Flash memory size mismatch detected. Expected 4MB, found 2MB!
Please select a proper value in your `sdkconfig.defaults` or via the `menuconfig` target!
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 0 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Compiling .pio/build/esp32dev/src/main.cpp.o
Generating LD script .pio/build/esp32dev/memory.ld
Compiling .pio/build/esp32dev/app_trace/app_trace.c.o
Compiling .pio/build/esp32dev/app_trace/app_trace_util.c.o
cc1: warning: command-line option '-Wno-overloaded-virtual=0' is valid for C++/ObjC++ but not for C
Compiling .pio/build/esp32dev/app_trace/host_file_io.c.o
cc1: warning: command-line option '-Wno-overloaded-virtual=0' is valid for C++/ObjC++ but not for C
cc1: warning: command-line option '-Wno-overloaded-virtual=0' is valid for C++/ObjC++ but not for C
```
EDIT: ah but this is with un-updated platforms... I'll try again

**Reply by crosswick** [reply]
same thing after updating the Espressif 32 platform:

```
Processing esp32dev (platform: espressif32; board: esp32dev; framework: espidf)
----------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.8.1) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-espidf @ 3.50300.0 (5.3.0) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.9.0 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 1.23800.240113 (2.38.0) 
 - toolchain-xtensa-esp-elf @ 13.2.0+20240530
Reading CMake configuration...
Generating assembly for certificate bundle...
Warning! Flash memory size mismatch detected. Expected 4MB, found 2MB!
Please select a proper value in your `sdkconfig.defaults` or via the `menuconfig` target!
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 0 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Compiling .pio/build/esp32dev/src/main.cpp.o
Generating LD script .pio/build/esp32dev/esp-idf/esp_system/ld/memory.ld.in
Generating LD script .pio/build/esp32dev/esp-idf/esp_system/ld/sections.ld.in
Compiling .pio/build/esp32dev/app_trace/app_trace.c.o
Generating LD script .pio/build/esp32dev/memory.ld
Compiling .pio/build/esp32dev/app_trace/app_trace_util.c.o
cc1: warning: command-line option '-Wno-overloaded-virtual=0' is valid for C++/ObjC++ but not for C
Compiling .pio/build/esp32dev/app_trace/host_file_io.c.o
cc1: warning: command-line option '-Wno-overloaded-virtual=0' is valid for C++/ObjC++ but not for C
cc1: warning: command-line option '-Wno-overloaded-virtual=0' is valid for C++/ObjC++ but not for C
```

**Reply by pschatzmann** [reply]
These are just warnings and the program will be compiled anyhow correctly...

**Reply by crosswick** [reply]
Oof, my bad! 

It does build indeed, I had cancelled the process before. 

Thank you very much!

**Reply by pschatzmann** [reply]
Maybe it is better to suppress all warnings:

build_flags = -w

**Comment by pschatzmann** [comment]
Here is my final conclusion: It seems that all warnings are treated as errors now because the following options are added automatically -Wall -Werror=all  

You can resolve the issue with the following entry in platformio.ini 

build_unflags = -Werror=all

This just leaves the warnings as such and does not treat them as errors.

**Reply by crosswick** [reply]
alright - do you mean also as alternative to 

`build_flags = -Wno-overloaded-virtual`

or is that still necessary for now?

**Reply by pschatzmann** [reply]
Yes and this is definitely the better alternative

**Reply by crosswick** [reply]
I see, great. I guess there probably is an equivalent configuration setting in ESP-IDF (without PlatformIO), I might look into that.

**Comment by MadMat76** [comment]
Hello @pschatzmann and @crosswick, 
I've just tried the same: using vs-code with esp-idf extension. I've started with an empty template, renamed main.c to main.cpp and created a component directory where I've cloned the arduino-audio-tools into it. The CMakeLists.txt in the audio-tools component already defines all the stuff correctly (idf_component_register and target_compile_options with ESP32-CMAKE. So no edit required. But...

There are other dependencies to arduino-esp32 core: AudioTools.h -> AudioConfig.h -> wants to include esp32-hal-log.h. Which is an arduino file. Do you have to add these also to NoArduino.h? Are there also other dependencies? Has anyone succeeded with no-arduino in esp-idf? If not, I'll try to do it the plattformio way and include arduino as component into my idf-project too.

**Reply by MadMat76** [reply]
an btw, thanks @pschatzmann for this great work! I've started with Arduino-IDE and enjoyed the easyness of doing audio-stuff with your library. Because of some low-power options, that are not available within arduino-IDE, I switched to esp-idf in vscode. There, I used the ESP-ADF framework. Gosh - that was complicated. The simplest examples were 500 lines of code and if you don't own one of the audio-kit boards, you have to figure out how to use a simple custom board. And there is no udpstream or espnow-stream. Only a (complicated) http-stream. So thats why I really want to switch back to your library.

**Reply by pschatzmann** [reply]
This functionality is not active by default: I committed a correction to include this only for Arduino

**Reply by MadMat76** [reply]
Wow, that was a quick fix. Yes it compiled further - but the next error came:
C:/Users/matth/audio-tools-idf/components/arduino-audio-tools/src/AudioTools/CoreAudio/AudioHttp/URLStream.h:7:11: fatal error: Client.h: No such file or directory
    7 | # include <Client.h>

**Reply by pschatzmann** [reply]
URLStream depends on the Arduino Networking API!
Maybe you can try Arduino as Component szenario...

**Reply by MadMat76** [reply]
Thanks, I'll try that way and post here, if I succeed.
